RandBLAS/dense.hh

```diff
@@ -1,27 +1,189 @@
+    RNGState next_state;
+    const bool own_memory = true;
+    /////////////////////////////////////////////////////////////////////
+    //
+    //      Properties specific to dense sketching operators
+    //
+    /////////////////////////////////////////////////////////////////////
+
+    T *buff = nullptr;
+    bool filled = false;
+    bool persistent = true;
+    const blas::Layout layout = blas::Layout::ColMajor;
+
+    /////////////////////////////////////////////////////////////////////
+    //
+    //      Member functions must directly relate to memory management.
+    //
+    /////////////////////////////////////////////////////////////////////
+
+    //  Elementary constructor: needs an implementation
+    DenseSkOp(
+        DenseDist dist_,
+        const RNGState &state_,
+        T *buff_,
+        bool filled_,
+        bool persistent_,
+        blas::Layout layout_
+    );
+
+    //  Convenience constructor (a wrapper)
+    DenseSkOp(
+        DenseDist dist,
+        uint32_t ctr_offset,
+        uint32_t key,
+        T *buff,
+        bool filled,
+        bool persistent,
+        blas::Layout layout
+    ) : DenseSkOp(dist, RNGState{ctr_offset, key}, buff, filled, persistent, layout) {};
+
+    //  Convenience constructor (a wrapper)
+    DenseSkOp(
+        DenseDistName family,
+        int64_t n_rows,
+        int64_t n_cols,
+        uint32_t key,
+        uint32_t ctr_offset,
+        T *buff,
+        bool filled,
+        bool persistent,
+        blas::Layout layout
+    ) : DenseSkOp(DenseDist{family, n_rows, n_cols}, RNGState{key, ctr_offset},
+        buff, filled, persistent, layout) {};
+
+    // Destructor
+    ~DenseSkOp();
+};
+
+template <typename T>
+DenseSkOp<T>::DenseSkOp(
+    DenseDist dist_,
+    const RNGState &state_,
+    T *buff_,           
+    bool filled_,       
+    bool persistent_,   
+    blas::Layout layout_ 
+) : // variable definitions
+    dist(dist_),
+    seed_state(state_),
+    next_state(),
+    own_memory(!buff_),
+    buff(buff_),
+    filled(filled_),
+    persistent(persistent_),
+    layout(layout_)
+{   // Initialization logic
+    //
+    //      own_memory is a bool that's true iff buff_ is nullptr.
+    //
+    if (this->own_memory) {
+        randblas_require(!this->filled);
+        // We own the rights to the memory, and the memory
+        // hasn't been allocated, so there's no way that the memory exists yet.
+    } else {
+        randblas_require(this->persistent);
+        // If the user gives us any memory to work with, then we cannot take
+        // responsibility for deallocating on exit from LSKGE3 / RSKGE3.
+    }
+}
+
+template <typename T>
+DenseSkOp<T>::~DenseSkOp() {
+    if (this->own_memory) {
+        delete [] this->buff;
+    }
+}
+
+
+
+
+
+
+
+
+
+
 template <typename T, typename T_gen>
 static RNGState gen_unif(
     int64_t n_rows,
     int64_t n_cols,
     T* mat,
-    RNGState state
+    const RNGState &state
```

Comment:
Originally I passed by value so that it didn't matter if we made changes to ``RNGState``. My guess is that this change of passing by a const reference is slightly more efficient. Is that reading correct?

Reply:
that makes sense. passing by reference to const is more efficient because we skip the deep copy made by the copy constructor. because it is a ref to const compiler will ensure we do not modify it. my 2 cents  : when passing an object by value intentionally we should add a comment that we're doing it on purpose because it is usually a mistake to pass by value and someone like me might change it leading to a bug.

Reply:
Sounds good; I'm on board with explicitly acknowledging passing by value whenever we do need it (which I imagine will be rare).